### ISSUE TITLE
`@NotEmpty`/`@NotNull` container element validators run against empty `Optional`

### ISSUE THREAD

**Issue by krnhotwings** [quarkusio/quarkus#41262]
### Describe the bug

```java
@GET
@Produces(MediaType.TEXT_PLAIN)
public String hello(@RestQuery Optional<@NotEmpty String> param) {
    return "Hello from Quarkus REST";
}
```

if `?param` is not present, the validator will still run against the empty `Optional` and return a validation error instead of the expected `200` response.

### Expected behavior

The validator should only run when the element is present. For example, `Optional<@Email String> param` will return 200 when `param` is not present but will validate `param` when present. Other containers like `List<@NotEmpty String> param` also work as expected. The issue seems to be explicitly `Optional` + `@NotEmpty`/`@NotNull`. I'm not sure if other single-element container types are also affected.

### Actual behavior

`Optional<@NotEmpty String>` and `@NotNull` will run their validators against the empty element and return:

```
ViolationReport{title='Constraint Violation', status=400, violations=[Violation{field='hello.param', message='must not be empty'}]}
```

### How to Reproduce?

1. Create new project and add `hibernate-validator` extension
2. See above example REST resource

### Output of `uname -a` or `ver`

6.9.4-200.fc40.x86_64

### Output of `java -version`

Temurin-21.0.3+9

### Quarkus version or git rev

3.11.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

gradle 8.8

**Comment by krnhotwings** [comment]
I should add that the bug essentially treats `Optional<@NotEmpty String> param` as `@NotEmpty String param`

**Comment by krnhotwings** [comment]
Ah, I should also note that this affects beans:

```java
@GET
@Produces(MediaType.TEXT_PLAIN)
public String hello(@Valid Bean bean) {
    return "Hello from Quarkus REST";
}
```

```java
public class Bean {
    public Optional<@NotEmpty String> param;
}
```

**Comment by krnhotwings** [comment]
Ah, I should also mention that this also affects validation against beans where optionals are initialized to empty:

```java
@POST
public String hello(@Valid MyBean bean) {
    return bean;
}
```

```java
public class MyBean {
    public Optional<@NotEmpty String> param = Optional.empty();
}
```

The above will return a validation error against `MyBean.param`

**Comment by marko-bekhta** [comment]
hey @krnhotwings 

The problem you've encountered is caused by the value extractor for optional.. and also the `@NotEmpty` constraint. Unlike most of the other constraints, it does not allow `null` values. 

looking at the spec here: https://jakarta.ee/specifications/bean-validation/3.0/jakarta-bean-validation-spec-3.0.html#valueextractordefinition-builtinvalueextractors

> java.util.Optional; value() must be invoked, passing null as node name and passing the contained object as value or null if none is present

the behaviour you've described matches these expectations. If we'd want to change that -- that should be a change in the spec (feel free to start a discussion here https://github.com/jakartaee/validation/issues). 

Now to how you can address your case to make things work as you'd like?

### Override the value extractor for the optional

1. Create an implementation like:
```java
public class OptionalValueExtractor implements ValueExtractor<Optional<@ExtractedValue ?>> {

	@Override
	public void extractValues(Optional<?> originalValue, ValueExtractor.ValueReceiver receiver) {
		// only pass an extracted value if something is present:
		if ( originalValue.isPresent() ) {
			receiver.value(null, originalValue.get());
		}
	}
}
```
2. Register this implementation via a `META-INF/services/jakarta.validation.valueextraction.ValueExtractor`
```
some.package.where.the.impl.is.OptionalValueExtractor
```

This will override the builtin implementation of the value extractor and things will work as you'd wanted.

### Do not use constraints that fail on `null` values on optional

As mentioned above, per spec if optional is empty -- null is passed for the validation. Soooo just don't use a constraint that fails on `null` values. You can create your own `@NullOrNotEmpty` and provide the implementation for the `ConstraintValidator` that would allow null values. Here's a blog post on how to add your custom constraints if you'd go this route: https://in.relation.to/2017/03/02/adding-custom-constraint-definitions-via-the-java-service-loader/ .

**Comment by krnhotwings** [comment]
@marko-bekhta Ah, sorry, I realize I had mistyped. I had meant to mention `@NotBlank` rather than `@NotEmpty`, but it appears the outcome is the same.

The issue is that if one has an optional string parameter (query param, form param, etc.) whose value should be not-blank if present, using `Optional<@NotBlank String>` will return a constraint violation error even if the param isn't present in the request.

This seems unusual since other validators like `@Email` will work as intended; that is: `Optional<@Email String>` will do nothing if the parameter isn't present, but it will validate the input value to check if the string is an email address if the parameter is present.

Given the above, would it still be the case that users would need to create a custom constraint for this scenario? Or is there a bug with the behavior of `@NotBlank`?

**Comment by marko-bekhta** [comment]
Hey, no worries 😃.

Right ... let me maybe rephrase it a bit 😃. I see and understand your point, but according to the specification https://jakarta.ee/specifications/bean-validation/3.0/jakarta-bean-validation-spec-3.0.html#valueextractordefinition-builtinvalueextractors

> java.util.Optional; value() must be invoked, passing null as node name and passing the contained object as value or null if none is present

which essentially means, if an optional is empty -- pass `null` to constraint validators. And Hibernate Validator must follow the spec... Now if you think that the specification should be updated, then a discussion on the specification side should be started (here https://github.com/jakartaee/validation/issues). 

> This seems unusual since other validators like @Email will work as intended; that is: Optional<@Email String> will do nothing if the parameter isn't present, but it will validate the input value to check if the string is an email address if the parameter is present.

right, so most validators would have something like:
```
if ( value == null || value.length() == 0 ) {
	return true;
}
```
as the very first check (e.g. for Email: https://github.com/hibernate/hibernate-validator/blob/c062b09e5f949e2d60357052c5bbf083a3b90e08/engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/AbstractEmailValidator.java#L51) 
This is done on purpose (considering null values as valid by most of the constraints), as there is a `@NotNull` constraint that can be added to those properties that shouldn't be null. 
And what happens if you have something like `Optional<@Email String> opt = Optional.empty()` is that a `null` value is passed to an email validator, but because of that, if statement it is considered a valid value 🙈 😃.

> Or is there a bug with the behavior of @NotBlank?

as per specification:

> The annotated element must not be null and must contain at least one non-whitespace character.

This means that `@NotBlank` has an expected behaviour. 

I understand that the way an extractor for an Optional behaves makes it look unusual, as you've said. I cannot remember why it was decided to do so back then, but I think there was some reasoning behind it..

At this point, I think the best option for you is to go with the `Override the value extractor for the optional` approach. And if you eventually want to get rid of the custom extractor, start that discussion on the Jakarta spec side.

**Comment by krnhotwings** [comment]
@marko-bekhta Thanks for all of the clarification; much appreciated.

I'll go with your suggestion to override the value extractor and look into bringing this up in the linked jakarta repo. Thanks!
